src/main/java/edu/immutables/list/impl/Cons.java

```diff
@@ -0,0 +1,160 @@
+package edu.immutables.list.impl;
+
+import java.util.Iterator;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+
+import static edu.immutables.list.impl.ImmutableList.reverse;
+
+public class Cons<T> implements ImmutableList<T> {
+    private final T head;
+    private final ImmutableList<T> tail;
+    private final int size;
+
+    Cons(T head) {
+        this(head, new Empty<>());
+    }
+
+    Cons(T head, ImmutableList<T> tail) {
+        this.head = head;
+        this.tail = tail;
+        this.size = 1 + tail.size();
+    }
+
+    @Override
+    public T get(int ind) {
+        if (ind == 0)
+            return getHead();
+        return tail.get(ind - 1);
+    }
+
+    @Override
+    public T getHead() {
+        return head;
+    }
+
+    @Override
+    public ImmutableList<T> getTail() {
+        return tail;
+    }
+
+    @Override
+    public ImmutableList<T> addHead(T t) {
+        return new Cons<>(t, this);
+    }
+
+    @Override
+    public ImmutableList<T> removeAt(int ind) {
+        if (ind == 0)
+            return tail;
+        return new Cons<>(head, tail.removeAt(ind - 1));
+    }
+
+    @Override
+    public ImmutableList<T> remove(T t) {
+        if (head.equals(t))
+            return tail;
+        return new Cons<>(head, tail.remove(t));
+    }
+
+    @Override
+    public <U> ImmutableList<U> map(Function<? super T, ? extends U> fn) {
+        return new Cons<>(fn.apply(head), tail.map(fn));
+    }
+
+    @Override
+    public <U> ImmutableList<U> flatMap(final Function<? super T, ImmutableList<U>> f) {
+        return f.apply(head).append(tail.flatMap(f));
+    }
+
+    @Override
+    public ImmutableList<T> takeWhile(Predicate<? super T> p) {
+        if (p.test(head)) {
+            return new Cons<>(head, tail.takeWhile(p));
+        }
+        return new Empty<>();
+    }
+
+    @Override
+    public ImmutableList<T> dropWhile(Predicate<? super T> p) {
+        return reverse(reverse(this).takeWhile(p.negate()));
+    }
+
+    @Override
+    public ImmutableList<T> take(int length) {
+        if (length == 0)
+            return new Empty<> ();
+        return new Cons<>(head, tail.take(length - 1));
+    }
+
+    @Override
+    public ImmutableList<T> drop(int length) {
+        return null;
+    }
+
+    @Override
+    public ImmutableList<T> append(ImmutableList<T> t) {
+        if (size == 1)
+            return new Cons<>(head, t);
+        return new Cons<>(head, tail.append(t));
+    }
```

Comment:
you don't have the if condition, simply return `new Cons<>(head, tail.append())` append implementation in `Empty` will end the recursive calls.